### ISSUE TITLE
Arrays of tuples aren't supported

### ISSUE THREAD

**Issue by reinerp** [LeventErkok/sbv#339]
I'd like to use arrays of tuples, e.g.:

`type MyArray = SFunArray SInt8 (SInt8, SInt8)`

While this type is indeed an instance of `SymArray`, the `writeArray` and `readArray` functions in that instance have type signatures which make them unusable:

`readArray :: MyArray -> SBV SInt8 -> SBV (SInt8, SInt8)
writeArray :: MyArray -> SBV SInt8 -> SBV (SInt8, SInt8) -> MyArray`

In particular, the `SBV (SInt8, SInt8)` part of the type signatures seems to be useless: there appears to be no way to extract the first `SInt8` element of an `SBV (SInt8, SInt8)`, or similarly to extract the second element. My current workaround is to roll my own functional arrays:

`type MyArray = SInt8 -> (SInt8, SInt8)`

**Comment by LeventErkok** [comment]
You're correct, and the behavior is intentional I'm afraid. Here's the typedef from SBV:

https://github.com/LeventErkok/sbv/blob/3b2e76fddc7d6a741996ed87cd0ef53c46607554/Data/SBV/Core/Data.hs#L474

Note that the symbolic part is automatically added, and can only be those things that can be treated symbolically by SMT-Lib. Tuples are not part of this schema.

Note that the expansion of your `MyArray` would be:

    SFunArray (SBV (SBV Int8)) (SBV (SBV Int8, SBV Int8))

which is not an array type SBV supports. But even if you said:

     type MyArray = SFunArray Int8 (Int8, Int8)

that wouldn't work either, since the expansion would be:

    SFunArray (SBV Int8) (SBV (Int8, Int8))

and as you observed, you can't really do much with `SBV (Int8, Int8)`.

The underlying reason for this is that SMT-Lib arrays are necessarily flat: http://smtlib.cs.uiowa.edu/logics-all.shtml#QF_AUFBV states:

>   Closed quantifier-free formulas built over an arbitrary expansion of the
>   FixedSizeBitVectors and ArraysEx signatures with free sort and function
>   symbols, but with the restriction that all array terms have sort of the 
>   form (Array (_ BitVec i) (_ BitVec j)) for some i, j > 0.

In more powerful logics you can have arrays of arbitrary elements, but not with bit-vectors. (Of course, different solvers might relax this requirement.)

Long story short, you are correct the array interface is less than useful when you want to store multiple values in every index. It's even less useful when you have multi-dimensional arrays. The solution is to essentially roll your own higher level types that boil down to the basics as supported by SMT-Lib and thus by SBV. Which is what I suspect what you are already doing.

Feel free to report further issues if you see them though. The array interface is one of the lesser tested areas in SBV and I'd love to hear feedback on it.

**Comment by reinerp** [comment]
Thanks Levent, good to know I'm not missing something.

I'd expect that the SBV library could itself expose the following definition of functional arrays:

`newtype SFunArray a b = SFunArray (a -> b)`

where `b` is any `Mergeable` type and `a` is any `EqSymbolic` type. That would add sufficient flexibility for all the cases I'm imagining at the moment.

**Comment by LeventErkok** [comment]
A more general type can indeed be added. My original goal was that `SFunArray` and `SArray` can be used interchangeably. (Former is the functional arrays, latter is the one that directly maps to SMT-Lib arrays, which are flat.) Hence the design was influenced by the common denominator.

But this was designed a long time ago, and honestly the "interchangeability" aspect didn't play much of a role in practice. To be honest, even use of arrays themselves didn't play much of a role, it's almost always better to just declare all the variables you need, unless you are truly dealing with memory-like structures.

Having said that, SBV is always open to patches and contributions. If you want to experiment and expand the definition of `SFunArray` to be more general, please contribute!

**Comment by LeventErkok** [comment]
@reinerp Do we have further action items on this ticket? I'd like to do a Christmas cleaning on the tickets.

**Comment by LeventErkok** [comment]
@reinerp I'm assuming this issue is now resolved? I'm planning to do a release in the next week or so and would like to make sure the concerns are addressed before doing so.

**Comment by reinerp** [comment]
This is resolved. I've been playing with some type class definitions which might address my need, but nothing needs to be done on this ticket. I'll send a patch if something comes of my work.

**Comment by LeventErkok** [comment]
Please do so. Thanks!
